### PR TITLE
feat: detect URL changes

### PR DESCRIPTION
**Pull request by bravecod**
### Summary

# URL Tracking Feature Changes

## Overview

This document outlines the comprehensive changes made to the URL tracking feature in the session-replay-browser package. The changes include converting dynamic imports to static imports, developing comprehensive test coverage, refactoring code for better maintainability, and adding detailed documentation.

## Table of Contents

- [1. Dynamic Import to Static Import Conversion](#1-dynamic-import-to-static-import-conversion)
- [2. Comprehensive Test Development](#2-comprehensive-test-development)
- [3. Code Refactoring](#3-code-refactoring)
- [4. Documentation Enhancements](#4-documentation-enhancements)
- [5. Performance and Maintainability Benefits](#5-performance-and-maintainability-benefits)
- [6. Testing Summary](#6-testing-summary)

---

### Changes Made

**File:** `packages/session-replay-browser/src/session-replay.ts`

#### Before
```typescript
const { URLTracker } = await import('./observers');
```

#### After
```typescript
import { URLTracker } from './observers';
```

### Implementation Details

- **Removed `async` keyword** from `initializeUrlTracking` method
- **Eliminated `await import('./observers')`** dynamic import
- **Updated method calls** to remove `void` wrapper since method is no longer async
- **Fixed shutdown tests** that were failing due to additional event listener removal

### Benefits

- ✅ **Improved startup performance** - No runtime module loading delay
- ✅ **Better bundle optimization** - Static analysis enables better tree shaking
- ✅ **Simplified code flow** - Synchronous initialization reduces complexity
- ✅ **Enhanced type safety** - Static imports provide better TypeScript support

---

### New Test File Created

**File:** `packages/session-replay-browser/test/url-tracker.test.ts`

#### Constructor Tests
- Default options initialization
- Custom options with UGC rules and polling

#### Lifecycle Management
- Start/stop tracking functionality
- Prevention of double initialization
- Graceful handling of missing global scope
- Event listener cleanup verification

#### URL Change Detection
- **pushState** navigation detection
- **replaceState** navigation detection  
- **popstate** event handling (back/forward navigation)
- Duplicate URL change prevention

#### UGC (User Generated Content) Filtering
- Application of filtering rules when configured
- Bypassing filtering when no rules exist
- Integration with `getPageUrl` helper function

#### Polling Mechanism
- Setup when enabled in configuration
- Disabled state verification
- Interval cleanup on stop
- Callback execution during polling

#### Configuration Updates
- Dynamic UGC filter rule updates
- Verification of updated rules application

#### Edge Case Handling
- Missing document title scenarios
- Undefined global scope conditions
- Missing location in global scope
- Callback handling when tracker is stopped
- Undefined document in global scope
- Null original methods in history patches

### Integration with Session Replay

**File:** `packages/session-replay-browser/test/session-replay.test.ts`

Added comprehensive integration tests:
- URL tracking initialization during SessionReplay init
- Configuration passing to URLTracker
- Event listener teardown including URLTracker cleanup
- Custom RRWeb event handling for URL changes
- Various configuration scenarios

---

### URLTracker Implementation Refactoring

**File:** `packages/session-replay-browser/src/observers/url-tracker.ts`

##### Constants Extraction
```typescript
const POLLING_INTERVAL_MS = 1000; // Eliminates magic numbers
```

##### Type Safety Enhancements
```typescript
type GlobalScope = NonNullable<ReturnType<typeof getGlobalScope>>;
```

##### Method Decomposition

**Before:** Large monolithic methods
**After:** Focused, single-responsibility methods

###### Setup Method Breakdown
- `setupUrlTracking()` → Main orchestration
- `storeOriginalHistoryMethods()` → Method storage
- `patchHistoryMethods()` → History API patching
- `setupEventListeners()` → Event listener setup
- `setupPolling()` → Polling configuration

###### Teardown Method Breakdown  
- `teardownUrlTracking()` → Main orchestration
- `restoreOriginalHistoryMethods()` → Method restoration
- `removeEventListeners()` → Event cleanup
- `clearPolling()` → Polling cleanup
- `resetState()` → State reset

##### Code Duplication Elimination

**Generic History Method Patching:**
```typescript
private createHistoryMethodPatch<T extends typeof history.pushState | typeof history.replaceState>(
  originalMethod: T | null
) {
  const emitUrlChange = this.emitUrlChange;
  return function (this: History, ...args: Parameters<T>) {
    const result = originalMethod?.apply(this, args);
    emitUrlChange();
    return result;
  };
}
```

##### Helper Method Addition
```typescript
private shouldApplyUgcFiltering(): boolean {
  return this.ugcFilterRules.length > 0;
}
```

#### Benefits of Refactoring

- ✅ **Improved maintainability** - Smaller, focused methods
- ✅ **Enhanced readability** - Clear separation of concerns
- ✅ **Better testability** - Individual methods can be tested in isolation
- ✅ **Reduced duplication** - Generic patching method eliminates code repetition
- ✅ **Type safety** - Eliminated unsafe `globalThis` usage

### Test Refactoring

**File:** `packages/session-replay-browser/test/url-tracker.test.ts`

#### Test Helper Functions Created
```typescript
// Test data factories
const createMockUgcFilterRules = () => [/* ... */];
const createMockGlobalScope = (overrides = {}) => ({ /* ... */ });

// Test utility functions  
const setUrlAndTitle = (url, title) => { /* ... */ };
const startTrackingAndClearCallback = () => { /* ... */ };
const mockGlobalScopeReturn = (scope) => { /* ... */ };
const getPopstateListener = () => { /* ... */ };
```

#### Benefits of Test Refactoring

- ✅ **Reduced code duplication** - Helper functions eliminate repetitive patterns
- ✅ **Improved readability** - Tests focus on business logic rather than setup
- ✅ **Easier maintenance** - Changes to test patterns only need updates in one place
- ✅ **Consistent patterns** - Standardized approach across all tests

---

### Comprehensive Code Documentation

**File:** `packages/session-replay-browser/src/observers/url-tracker.ts`

##### Class-Level Documentation
```typescript
/**
 * URLTracker monitors URL changes in single-page applications by:
 * 1. Patching browser history methods (pushState, replaceState)
 * 2. Listening for popstate events (back/forward navigation)  
 * 3. Optional polling as a fallback for edge cases
 * 
 * Features:
 * - Deduplication of identical URL changes
 * - UGC (User Generated Content) filtering for sensitive URLs
 * - Graceful handling of missing browser APIs
 * - Clean teardown with method restoration
 */
```

##### Method Documentation
- **JSDoc comments** for all public and private methods
- **Parameter descriptions** with types and purposes
- **Return value documentation** where applicable
- **Implementation rationale** explanations

##### Property Organization
- **Grouped related properties** with explanatory comments
- **Explained purpose** of each property group
- **Clarified state management** logic

##### Inline Comments
- **Early return explanations** 
- **Sequence of operations** in complex methods
- **Browser behavior explanations**
- **Edge case handling rationale**

#### Documentation Benefits

- ✅ **Faster developer onboarding** - New team members understand code quickly
- ✅ **Easier maintenance** - Implementation decisions are documented
- ✅ **Better debugging** - Edge cases and error handling are explained
- ✅ **Reduced cognitive load** - Developers don't need to reverse-engineer logic

---

#### Static Import Benefits
- **Faster startup time** - No runtime module resolution
- **Better bundle optimization** - Static analysis enables tree shaking
- **Reduced memory overhead** - No dynamic import machinery

#### Refactoring Benefits  
- **Smaller method complexity** - Easier for JavaScript engines to optimize
- **Better type inference** - TypeScript can optimize more effectively
- **Reduced object creation** - Reusable generic method patterns

#### Code Structure
- **Single Responsibility Principle** - Each method has one clear purpose
- **DRY Principle** - Code duplication eliminated through generic methods
- **Clear Separation of Concerns** - Setup, operation, and teardown are distinct

#### Testing
- **100% Code Coverage** - All branches, statements, functions, and lines covered
- **Comprehensive Edge Cases** - Handles various browser environment scenarios
- **Maintainable Test Suite** - Helper functions reduce duplication and improve clarity

#### Documentation
- **Self-Documenting Code** - Clear method names and comprehensive comments
- **Implementation Rationale** - Why decisions were made, not just what was done
- **API Documentation** - Clear parameter and return value descriptions

---

### Coverage Metrics

**Final Coverage Results:**
- **Statements**: 100% (98/98)
- **Branches**: 100% (36/36) 
- **Functions**: 100% (20/20)
- **Lines**: 100% (91/91)

#### URL Tracker Tests (`url-tracker.test.ts`)
- **24 comprehensive tests** covering all functionality
- **Organized test suites** by feature area
- **Edge case coverage** for error conditions
- **Integration verification** with helper utilities

#### Session Replay Integration Tests (`session-replay.test.ts`)
- **URL tracking initialization** during SessionReplay setup
- **Configuration propagation** testing
- **Event listener management** verification
- **Custom RRWeb event handling** validation

#### Before Optimization
- **25 tests** with some redundancy
- **97.22% branch coverage** (35/36 branches)
- **Overlapping test scenarios**

#### After Optimization  
- **24 focused tests** with eliminated redundancy
- **100% branch coverage** (36/36 branches)
- **Streamlined test scenarios** without coverage loss

### Edge Cases Covered

1. **Browser API Availability**
   - Missing global scope
   - Undefined location object
   - Missing document object

2. **Method State Management**
   - Null original history methods
   - Undefined callback scenarios
   - Double initialization prevention

3. **URL Change Scenarios**
   - Duplicate URL prevention
   - Missing document titles
   - Various navigation methods

4. **Configuration Edge Cases**
   - Empty UGC filter rules
   - Missing interaction config
   - Dynamic configuration updates

---

## Summary

The URL tracking feature has undergone comprehensive improvements across multiple dimensions:

### ✅ **Code Quality**
- Converted from dynamic to static imports
- Comprehensive refactoring with modern patterns
- 100% test coverage with optimized test suite
- Extensive documentation for maintainability

### ✅ **Performance**  
- Faster initialization with static imports
- Better bundle optimization opportunities
- Reduced runtime complexity

### ✅ **Maintainability**
- Clear separation of concerns
- Comprehensive documentation
- Robust error handling
- Extensive test coverage

### ✅ **Developer Experience**
- Well-documented API surface
- Clear implementation patterns
- Comprehensive test examples
- Easy-to-understand code structure

These changes establish a solid foundation for the URL tracking feature that is performant, maintainable, and thoroughly tested. 

### Checklist

* [ ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:

### CODE REVIEW

File: packages/session-replay-browser/src/observers/url-tracker.ts [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+  title: string;
+}
+
+/**
+ * Callback function type for URL change notifications
+ */
+export type URLChangeCallback = (event: URLChangeEvent) => void;
+
+// Constants
+const POLLING_INTERVAL_MS = 1000; // Check for URL changes every second when polling is enabled
```
Comment (bravecod) [inline packages/session-replay-browser/src/observers/url-tracker.ts]: make a config to allow users to update it on their end

File: packages/session-replay-browser/src/session-replay.ts [hunk 2]
```diff
@@ -98,9 +100,31 @@ export class SessionReplay implements AmplitudeSessionReplay {
...
+      ugcFilterRules: this.config.interactionConfig?.ugcFilterRules || [],
+      enablePolling: this.config.enableUrlChangePolling || false,
+    });
+
+    // Start tracking and handle URL changes
+    this.urlTracker.start((event) => {
+      void this.addCustomRRWebEvent(CustomRRwebEvent.URL_CHANGE, event);
+    });
+  };
+
```
Comment (bravecod) [inline packages/session-replay-browser/src/session-replay.ts]: add logic to prevent multiple instances

File: packages/session-replay-browser/src/constants.ts [hunk 3]
```diff
@@ -32,4 +32,5 @@ export enum CustomRRwebEvent {
   DEBUG_INFO = 'debug-info',
   FETCH_REQUEST = 'fetch-request',
   METADATA = 'metadata',
+  URL_CHANGE = 'url-change',
```
Comment (bravecod) [inline packages/session-replay-browser/src/constants.ts]: Can we reuse this metadata?

File: packages/session-replay-browser/src/observers/url-tracker.ts [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+   */
+  updateConfig(ugcFilterRules: UGCFilterRule[]): void {
+    this.ugcFilterRules = ugcFilterRules;
+  }
+
+  /**
+   * Emit a URL change event if the URL has actually changed
+   * Uses arrow function to preserve 'this' context when called from patched methods
+   */
+  private emitUrlChange = (): void => {
```
Comment (bravecod) [inline packages/session-replay-browser/src/observers/url-tracker.ts]: can we emit the screen size? (viewport)

Comment (bravecod) [reply packages/session-replay-browser/src/observers/url-tracker.ts]: image

File: packages/session-replay-browser/src/observers/url-tracker.ts [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+  viewportWidth: number;
+}
+
+/**
+ * Callback function type for URL change notifications
+ */
+export type URLChangeCallback = (event: URLChangeEvent) => void;
+
+// Constants
+const POLLING_INTERVAL_MS = 1000; // Check for URL changes every second when polling is enabled
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/observers/url-tracker.ts]: Can we make this a config instead?

**Review by lewgordon-amplitude — COMMENTED** [review]
Looks good, especially since we chatted through most of this. I just would want to add make sure things are changeable in config before we ship it.

File: packages/session-replay-browser/src/observers/url-tracker.ts [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    // Bail out if browser doesn't support location API
+    if (!globalScope || !globalScope.location) return;
+
+    const { innerHeight, innerWidth, location, document } = globalScope;
+
+    const currentUrl = location.href;
+    const currentTitle = document?.title || '';
+
+    // Only emit if URL actually changed to prevent duplicate events
+    if (currentUrl !== this.lastTrackedUrl) {
```
Comment (erccarls) [inline packages/session-replay-browser/src/observers/url-tracker.ts]: I am looking at a similar tracking concept downstream of this.    It is worth considering how some SPA / router frameworks could be integrated here: 

```
RRWeb's built-in Meta events fire only on full‐page navigations, so modern single-page apps (React Router, Vue Router, etc.) never emit them. The remedy is to instrument the History and Hash routing APIs in the recorder and emit a lightweight custom event each time the URL, document title, or referrer changes.```

Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/observers/url-tracker.ts]: @erccarls yeah I think the goal here is definitely to solve for those use cases. Or if it doesn't I think we need to iterate to solve for those.

File: packages/session-replay-browser/src/constants.ts [hunk 7]
```diff
@@ -32,4 +33,5 @@ export enum CustomRRwebEvent {
...
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/constants.ts]: Is this still needed with the plugin change?

File: packages/session-replay-browser/src/plugins/url-tracking-plugin.ts [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        }
+
+        // Apply UGC filtering if rules are provided, otherwise use original URL
+        const filteredUrl = ugcFilterRules.length > 0 ? getPageUrl(currentUrl, ugcFilterRules) : currentUrl;
+
+        return {
+          href: filteredUrl,
+          title: currentTitle,
+          viewportHeight: innerHeight,
+          viewportWidth: innerWidth,
```
Comment (jxiwang) [inline packages/session-replay-browser/src/plugins/url-tracking-plugin.ts:112]: Any additional information we might consider useful here? Some that come to mind are the query or timestamp.

File: packages/session-replay-browser/src/plugins/url-tracking-plugin.ts [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+  options: URLTrackingPluginOptions = {},
+): RecordPlugin<URLTrackingPluginOptions> {
+  return {
+    name: 'amplitude/url-tracking@1',
+    observer(cb, globalScope, pluginOptions?: URLTrackingPluginOptions) {
+      // Merge options with plugin-level options taking precedence over constructor options
+      const config = { ...options, ...pluginOptions };
+      const ugcFilterRules = config.ugcFilterRules || [];
+      const enablePolling = config.enablePolling || false;
+      const pollingInterval = config.pollingInterval || DEFAULT_URL_CHANGE_POLLING_INTERVAL;
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/plugins/url-tracking-plugin.ts]: Nit: Why not `??` here instead?

File: packages/session-replay-browser/src/plugins/url-tracking-plugin.ts (hunk 8 above)
Comment (cely404) [reply packages/session-replay-browser/src/plugins/url-tracking-plugin.ts:112]: You mean like query params?

File: packages/session-replay-browser/src/plugins/url-tracking-plugin.ts [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        // Return cleanup function to stop polling
+        return () => {
+          if (urlChangeInterval) {
+            globalScope.clearInterval(urlChangeInterval);
+          }
+        };
+      }
+
+      // History API mode: patch history methods and listen to popstate events
+      if (globalScope.history) {
```
Comment (cely404) [inline packages/session-replay-browser/src/plugins/url-tracking-plugin.ts:177]: Do we need to explicity set this as an if, else if?
In order to prevent this from doing the same thing in different ways

**Review by cely404 — APPROVED** [review]

File: packages/session-replay-browser/src/plugins/url-tracking-plugin.ts (hunk 10 above)
Comment (bravecod) [reply packages/session-replay-browser/src/plugins/url-tracking-plugin.ts:177]: sorry can you elaborate on this?

File: packages/session-replay-browser/src/observers/url-tracker.ts (hunk 6 above)
Comment (cely404) [reply packages/session-replay-browser/src/observers/url-tracker.ts]: @bravecod are there proposed changes to your current solution? 
If so, should we push back the test fest once more?

File: packages/session-replay-browser/src/plugins/url-tracking-plugin.ts (hunk 10 above)
Comment (cely404) [reply packages/session-replay-browser/src/plugins/url-tracking-plugin.ts:177]: As in should we have 

If(globalScope.history){}
else if(enablePolling){}

I might be reading this wrong but it looks like both blocks of code can execute and I think we want only one

Comment (bravecod) [reply packages/session-replay-browser/src/plugins/url-tracking-plugin.ts:177]: oh, it won't. It will exit since there's a return statement so it's optional to have an elseif
image

Comment (cely404) [reply packages/session-replay-browser/src/plugins/url-tracking-plugin.ts:177]: Ah mb, missed that

File: packages/session-replay-browser/src/observers/url-tracker.ts (hunk 6 above)
Comment (bravecod) [reply packages/session-replay-browser/src/observers/url-tracker.ts]: yeah, I have updated it so it captures all these
Scenario 1: History API Mode (Default)
Scenario 2: Polling Mode
Scenario 3: Hash Router

// 1. if explicitly enable polling → use polling
if (enablePolling) {
  // Use polling (covers everything)
} 
// 2. if polling not enabled → check history, if exist, use history
else if (globalScope.history) {
  // Use History API + hashchange (covers History API + hash routing)
  globalScope.addEventListener('popstate', emitUrlChange);
  globalScope.addEventListener('hashchange', hashChangeHandler);
} 
// 3. if not, then the framework is probably using hash router → do hash
else {
  // Fallback: just hashchange (for pure hash routing)
  globalScope.addEventListener('hashchange', hashChangeHandler);
}

**Comment by jpollock-ampl** [comment]
bugbot run

**Comment by bravecod** [comment]
> ### Bug: URL Tracker Initialization Overwrites Without Cleanup
> The `initializeUrlTracking` method creates a new `URLTracker` instance without stopping the existing one. Repeated calls to `init()` or `teardownEventListeners(false)` result in multiple `URLTracker` instances running simultaneously, leading to duplicate URL change events, memory leaks, and potential incorrect behavior. The method should stop the existing `URLTracker` before creating a new instance.
> 
> `packages/session-replay-browser/src/session-replay.ts#L112-L126`
> [Fix in Cursor](https://cursor.com/open?data=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImJ1Z2JvdC12MSJ9.eyJ2ZXJzaW9uIjoxLCJ0eXBlIjoiQlVHQk9UX0ZJWF9JTl9DVVJTT1IiLCJkYXRhIjp7InJlZGlzS2V5IjoiYnVnYm90OjdlMWZjNDQ4LTM0YTEtNDRiMS05MmQ4LWNjYjRjNDg3MTFkZSIsImVuY3J5cHRpb25LZXkiOiJpazBsYmJHRGstSWhlYVJpZVhyaU1SNEpJSWpWeFhtbWIxYUhiQkFCV29ZIiwiYnJhbmNoIjoidXJsLXRyYWNraW5nIn0sImlhdCI6MTc1MTk5MzEyMiwiZXhwIjoxNzUyNTk3OTIyfQ.WAa5AgyY1CQT1rdFtbYdCdMM6E6OwjTD6rOZwKAm17GtjAEBLSuBGIhWN3T3qu7k0JmGdPmGRMuBz1cmgnjtPOZvD4s_mzLOWikLIesDS5GbW2f_4tqKHgN-vxcnl2scg68A9Eic4mferDpJLhsMgS6J1TcZaoo3sG69hOSppgOYa2x4Zjzq2SzyUtYiYppJy_OBo33txNkuq9pDfR-_LKHchDyebiH8Ro1RcfO0k4QCEdKlTGykp0FW-HOnjqUSWvkgASK7uVvyck1pKa6XKQsUG-T18g9KYKahMaJusHjQwyBiTnRyR-kTpZZxW9Cjla45rBM0ZWLCpAzacXxldg) • [Fix in Web](https://cursor.com/agents?data=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImJ1Z2JvdC12MSJ9.eyJ2ZXJzaW9uIjoxLCJ0eXBlIjoiQlVHQk9UX0ZJWF9JTl9XRUIiLCJkYXRhIjp7InJlZGlzS2V5IjoiYnVnYm90OjdlMWZjNDQ4LTM0YTEtNDRiMS05MmQ4LWNjYjRjNDg3MTFkZSIsImVuY3J5cHRpb25LZXkiOiJpazBsYmJHRGstSWhlYVJpZVhyaU1SNEpJSWpWeFhtbWIxYUhiQkFCV29ZIiwiYnJhbmNoIjoidXJsLXRyYWNraW5nIiwicmVwb093bmVyIjoiYW1wbGl0dWRlIiwicmVwb05hbWUiOiJBbXBsaXR1ZGUtVHlwZVNjcmlwdCIsInByTnVtYmVyIjoxMTkyLCJjb21taXRTaGEiOiI0N2UxZWRmMTliZjU5ZjExYTcxZmVkYmVmMDU5ODNiZGIxNDIzYTQ0In0sImlhdCI6MTc1MTk5MzEyMiwiZXhwIjoxNzUyNTk3OTIyfQ.LhorcrWXp9FVeMU4y8CtusB_BmwcO404zpxsTiUukuLfiLxPJ_EMKMGQhOl9_mnMOdU6RELxCBQOXFTa82ywW1qutwCUkgOYqOQrQYFBQLxgswLAzVWdArtAFWL8lwo-cYL5OddLEgQmCa8HZalSLhhYASe10-ZtYbXwjckkMrDlRCt_nZA_-wOthogrXk63GasaZnNpjv9chP42_iPUp4_XjYKDeAWeD3FLkvACx3sqMfEffc9svDJnet2XHQ3DGx_Lgn_CSCcjSWFuxdc8GWx8_Xd2QAfgcbipHvbHpnRsCwAerwaKsckkNM8lTozqndSp9RPnY7QZc0Mc5DSeoQ)
> 
> _Comment `bugbot run` to trigger another review on this PR_ _Was this report helpful? Give feedback by reacting with 👍 or 👎_

The reported bug has been resolved through the architectural changes. The current implementation:
✅ Creates only one URL tracking plugin per recording session
✅ Properly cleans up previous instances before creating new ones
✅ Uses rrweb's plugin lifecycle management for proper cleanup
✅ Prevents duplicate event listeners and memory leaks
The bug report can be considered resolved and is no longer applicable to the current codebase.

**Comment by bravecod** [comment]
bugbot run

**Review by erccarls — APPROVED** [review]
